rerun/launcher.go

```diff
@@ -0,0 +1,216 @@
+	log       *logp.Logger
+	ctx       context.Context
+	latest    *config.C
+	beat      *beat.Beat
+	creator   beat.Creator
+	validator Validator
+}
+
+type Reloader interface {
+	Channel() <-chan *config.C
+}
+
+type Validator interface {
+	Validate(*config.C) error
+}
+
+func NewLauncher(ctx context.Context,
+	log *logp.Logger,
+	reloader Reloader,
+	validator Validator,
+	bt *beat.Beat,
+	creator beat.Creator,
+	cfg *config.C) (*launcher, error) {
+	s := &launcher{
+		wg:        sync.WaitGroup{},
+		ctx:       ctx,
+		log:       log,
+		reloader:  reloader,
+		validator: validator,
+		beaterErr: make(chan error, 1),
+		creator:   creator,
+		latest:    cfg,
+	}
+
+	return s, nil
+}
+
+func (s *launcher) Run(b *beat.Beat) error {
+	// Configure the beats Manager to start after all the reloadable hooks are initialized
+	// and shutdown when the function returns.
+	s.beat = b
+	if err := b.Manager.Start(); err != nil {
+		return err
+	}
+	defer b.Manager.Stop()
+
+	// Wait for Fleet-side reconfiguration only if beater is running in Agent-managed mode.
+	if b.Manager.Enabled() {
+		s.log.Infof("Waiting for initial reconfiguration from Fleet server...")
+		update, err := s.reconfigureWait(reconfigureWaitTimeout)
+		if err != nil {
+			s.log.Errorf("Failed while waiting for initial reconfiguraiton from Fleet server: %v", err)
+			return err
+		}
+
+		if err := s.configUpdate(update); err != nil {
+			return fmt.Errorf("failed to update with initial reconfiguration from Fleet server: %w", err)
+		}
+	}
+
+	return s.run()
+}
+
+func (s *launcher) run() error {
+	err := s.runBeater()
+	if err != nil {
+		return err
+	}
+
+	err = s.waitForUpdates()
+	s.log.Error("Beater starter is stopping: %w", err)
+	return err
+}
+
+func (s *launcher) Stop() {
+	s.stopBeater()
+}
+
+func (s *launcher) runBeater() error {
+	beater, err := s.creator(s.beat, s.latest)
+	if err != nil {
+		return fmt.Errorf("Could not create beater: %w", err)
+	}
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		s.beaterErr <- beater.Run(s.beat)
+	}()
+
+	s.beater = beater
+	return nil
+}
+
+func (s *launcher) stopBeater() {
+	s.beater.Stop()
+	s.wg.Wait()
+}
+
+func (s *launcher) waitForUpdates() error {
+	for {
+		select {
+		case <-s.ctx.Done():
+			s.stopBeater()
+			return nil
+
+		case err := <-s.beaterErr:
+			if err != nil {
+				return fmt.Errorf("Beater returned an error:  %w", err)
+			}
+
+		case update, ok := <-s.reloader.Channel():
+			if !ok {
+				return errors.New("Reloader channel closed")
+			}
+
+			go func() {
+				if err := s.configUpdate(update); err != nil {
+					s.log.Errorf("Failed to update beater config: %v", err)
+				}
+			}()
+		}
+	}
+}
+
+// configUpdate applies incoming reconfiguration from the Fleet server to the beater config,
+// and recreate the beater with the new values.
+func (s *launcher) configUpdate(update *config.C) error {
+	s.log.Info("Got config update")
+
+	err := s.latest.MergeWithOpts(update, ucfg.ReplaceArrValues)
```

Comment:
Since we're doing a merge and not a replace, given that the incoming configuration is only partial, maybe the deleted tests from the earlier "merge configuration" logic are still useful as a double check?

Unrelated, did you explore having the integration send the whole configuration?

Reply:
> Regarding the first comment, not sure what you meant.

I was talking about some deleted tests, but I see now that they are replaced by analogous tests. Nevermind!